src/js/osweb/items/loop.js

```diff
@@ -182,85 +213,68 @@ export default class Loop extends Item {
 
   /** Implements the prepare phase of an item. */
   prepare () {
-    //  First generate a list of cycle numbers
-    this._cycles = []
-    this._index = 0
-
-    // Walk through all complete repeats
-    var wholeRepeats = Math.floor(this.vars.get('repeat'))
-    for (let j = 0; j < wholeRepeats; j++) {
-      for (let i = 0; i < this.vars.cycles; i++) {
-        this._cycles.push(i)
-      }
-    }
-
-    // Add the leftover repeats.
-    const partialRepeats = this.vars.get('repeat') - wholeRepeats
-    if (partialRepeats > 0) {
-      const allCycles = Array.apply(null, {
-        length: this.vars.cycles
-      }).map(Number.call, Number)
-      const remainder = Math.floor(this.vars.cycles * partialRepeats)
-      for (let i = 0; i < remainder; i++) {
-        // Calculate random position.
-        const position = Math.floor(Math.random() * allCycles.length)
-        // Add position to cycles.
-        this._cycles.push(position)
-        // Remove position from array.
-        allCycles.splice(position, 1)
-      }
-    }
-
-    // Randomize the list if necessary.
-    if (this.vars.get('order') === 'random') {
-      this._cycles = shuffle(this._cycles)
-    } else {
-      const skipVal = this.vars.get('skip')
-      // In sequential order, the offset and the skip are relevant.
-      if (this._cycles.length < skipVal) {
-        this.experiment._runner._debugger.addError('The value of skip is too high in loop item. ' +
-          'You cannot skip more cycles than there are in: ' + this.name)
-      } else {
-        if (this.vars.get('offset') === 'yes') {
-          // Get the skip elements.
-          const skip = this._cycles.slice(0, skipVal)
-
-          // Remove the skip elements from the original location.
-          this._cycles = this._cycles.slice(skipVal)
-
-          // Add the skip element to the end.
-          this._cycles = this._cycles.concat(skip)
-        } else {
-          this._cycles = this._cycles.slice(skipVal)
-        }
-      }
-    }
-
-    this._keyboard = new Keyboard(this.experiment)
 
     // Make sure the item to run exists.
     if (this.experiment.items._items[this.vars.item] === 'undefined') {
-      this.experiment._runner._debugger.addError('Could not find an item which is called by loop item: ' +
+      this._runner._debugger.addError('Could not find an item which is called by loop item: ' +
       this.name + ' (' + this.vars.item + ')')
     }
-
     super.prepare()
-
     this.set_item_onset()
   }
 
   /** Implements the run phase of an item. */
   run () {
     super.run()
 
+    if (!this._initialized) {
+      // Randomize the list if necessary.
+      if (this.vars.get('order') === 'random') {
+        this._cycles = shuffle(this._cycles)
+      }
+

```

Comment:
In the original code, there is also something done depending on a `constraints` variable. The purpose of this was not entirely clear to me. Should this also be done here for osweb, or is this datamatrix specific?

Reply:
The constraints refer to the pseudorandomization functionality, which hasn't been implemented in OSWeb yet.

- https://osdoc.cogsci.nl/3.3/manual/structure/loop/#pseudorandomization

Reply:
So we can leave this as is for now?

Reply:
Yes